### DISCUSSION TITLE
Retry on 5.x?

### DISCUSSION THREAD

**Discussion by bvitale** [panva/openid-client#450] [Q&A] [answered]
We were using the `retry` option on 4.x, with a number (retry count). What is the default retry behavior in 5.x?

**Comment by panva** [comment, marked as answer]
No retry in 5.x

**Reply by bvitale** [reply]
Ok so we'd be handling any network errors ourselves (for example timeout exceeded) and retrying then? Just making sure I'm not missing something obvious ..

**Reply by panva** [reply]
If retrying failed in 4.x you'd have to do same. Only in v5.x you're dealing with node's https module errors and not got.

**Reply by paulius-valiunas** [reply]
why?

**Reply by panva** [reply]
Because keeping up with `got`'s breaking changes and release cadence became a burden. Sticking with node's http(s) responses feels more native for a Node.js runtime module like `openid-client`.

**Reply by paulius-valiunas** [reply]
would you be open to re-implementing some retry logic in your package?

**Reply by panva** [reply]
not really, https://www.npmjs.com/search?q=retry should let you find a suitable retry orchestration library if you need it.

**Reply by paulius-valiunas** [reply]
that's not my cup of tea, thank you. There is a reason libraries like `got` exist: you shouldn't just blindly retry every failed request, there are specific HTTP status codes and connection issues that should be retried, and others that shouldn't. No one is going to write and maintain a decent library that wraps around existing code and tries to understand its errors for adding retry logic because:

1. the original code should already be using `got` or something similar
2. the original code might have its own error handling in place which would make it hard to detect which errors should be retried
3. maintainers of the original code are the ones to know best which of their requests should be retried and which shouldn't
4. the original code might have several consecutive network requests in one method, which would prevent said library from wrapping each request separately without having to monkey-patch the network stack, which would be gross to say the least.

If you think retry logic is unnecessary, sure, you're entitled to your opinion and I'm not asking you to change it. But there is no good external solution to this problem and there will never be one.

**Reply by paulius-valiunas** [reply]
by the way, since Node's http library still follows the ancient callback soup API style, there's no way for me to handle any errors coming from it. I'm looking at your code and I don't see any error handling, so what happens if there's a DNS error during a request? I suspect the application will simply crash due to an unhandled exception, but because the exception occurred in a scope I can't reach, there's nothing I can do about it. Of course I would have to try it out to confirm this, but it doesn't seem like there's a decent way to handle http exceptions with this package.

**Reply by panva** [reply]
> there's no way for me to handle any errors coming from it

> I'm looking at your code and I don't see any error handling, so what happens if there's a DNS error during a request?

> Of course I would have to try it out to confirm this, but it doesn't seem like there's a decent way to handle http exceptions with this package.

It's not that hard to try. As one might expect, you get the network errors and exceptions through the Promise APIs as rejections.

```js
import * as assert from 'node:assert'
import { Issuer } from 'openid-client'

try {
  await Issuer.discover('https://foo-does-not-exist.com')
  assert(0)
} catch (err) {
  assert.strictEqual(err.code, 'ENOTFOUND')
  console.log('DNS lookup failed, i might wanna retry Issuer.discover')
}
```

**Reply by paulius-valiunas** [reply]
you win this round :) I still think the correct place to have retry logic would be in this package because of all the reasons I mentioned above. But I guess me and all your other users will waste our time trying to figure out what kind of errors http.request can throw (it's not documented anywhere) and implementing the same thing on our own.

**Reply by panva** [reply]
I don't feel like I'm winning anything, in fact I am thinking whether to reconsider my position or not.

There have been no indices to so far in the year or so that 5.x is out. The most prevailing network related error is a timeout, which is dealt with by, well, increasing the permitted timeout duration through available means.

In my deployments of openid-client i'm not seeing ENOTFOUND, ECONNREFUSED, or ECONNRESET that frequently (or at all) to make me care.

Then, "specific HTTP status codes" you mention but did not provide much insight into, do not have a prescribed handling in OAuth 2.0, if an HTTP status code has a prescribed behaviour in OAuth 2.0 it is built in, usually with a IANA registered error code which is then echoed in the accompanied OPError rejection. Unexpected HTTP Responses usually come in the form of RPError with a description, those were never handled through the `retry` got mechanism anyway.

> me and all your other users

If this was a problem for "all my other users" there would be way more related activity around this since the 5.x release a year ago, and by now close to 50% of the weekly npm downloads are for 5.x, with no retry.

I personally would choose to not hide network issues behind the API because most likely a frequent network issue is not something to hide, it's something to deal with on the network/host configuration layer, not in the software, hiding the fact that there's something rotting on the network.

**Reply by paulius-valiunas** [reply]
oh yeah, I said it half-jokingly. I don't set up new packages for testing very often so in order to save time I just went with what Node.js documentation had to say, which is pretty much nothing. Of course my assumptions with the limited information I had were inaccurate, so I had to acknowledge your victory here :)

The specific status codes and other errors I had in mind are documented here: https://github.com/sindresorhus/got/blob/main/documentation/7-retry.md#retry

For example, in our deployments I've seen a few timeouts even with limit as long as 10 seconds, as well as EAI_AGAIN errors that come from DNS queries. I do agree that ideally the network should be made more stable so these things don't happen, but we don't live in an ideal world and that is not in my control. I'm not saying these errors should be handled by default, but I think it's very reasonable to expect being able to configure some retries if desired.

I also wish we lived in a world where auth servers only sent responses that were explicitly specified in OAuth 2.0, but again, we don't live in a perfect world and that means we get to deal with all kinds of unexpected issues from a 500 response to a temporary DNS outage.

And the most annoying part of all this is that Node.js doesn't document their errors properly so I'll have to read through `got`'s code to figure out what errors they process and how.

**Reply by panva** [reply]
I am confident Node.js documents their errors. I'm not able to look for that at this time tho.

**Reply by paulius-valiunas** [reply]
Well sure, they have this: https://nodejs.org/api/errors.html#nodejs-error-codes
...but there's no telling which error codes apply to http requests and which don't. Also the http page says nothing at all about errors, other than the fact that you can subscribe to `error` events:
https://nodejs.org/api/http.html#class-httpclientrequest
https://nodejs.org/api/http.html#httprequesturl-options-callback

So I can either guess around or refer to a known good implementation, such as got.
